plugins/builds/index.js

```diff
@@ -293,83 +410,329 @@ exports.register = (server, options, next) => {
+                    //    joinList doesn't include C, so start A
+                    if (joinList.length === 0 || !joinListNames.includes(currentJobName)) {
+                        return createBuild(buildConfig);
+                    }
+
+                    return Promise.resolve().then(() => {
+                        if (!event.parentEventId) {
+                            return event.getBuilds();
+                        }
+
+                        // If parent event id, merge parent build status data and
+                        // rerun all builds in the path of the startFrom
+                        return eventFactory.get({ id: event.parentEventId })
+                            .then(parentEvent => parentEvent.getBuilds()
+                                .then(parentBuilds => removeDownstreamBuilds({
+                                    builds: parentBuilds,
+                                    startFrom: event.startFrom,
+                                    parentEvent
+                                }))
+                            )
+                            .then(upstreamBuilds => event.getBuilds()
+                                .then(builds => builds.concat(upstreamBuilds)));
+                    }).then(finishedBuilds => handleNextBuild({
+                        buildConfig,
+                        joinList,
+                        finishedBuilds,
+                        jobId: workflowGraph.nodes
+                            .find(node => node.name === trimJobName(nextJobName)).id
+                    }));
+                }));
+            });
+        }
+
+        // New implementation that allows external join
+        const pipelineFactory = server.root.app.pipelineFactory;
+        const event = await eventFactory.get({ id: build.eventId });
+        const workflowGraph = event.workflowGraph;
+        const nextJobs = workflowParser.getNextJobs(workflowGraph,
+            { trigger: currentJobName, chainPR: pipeline.chainPR });
+        // Create a join object like: {A:[B,C], D:[B,F]} where [B,C] join on A, [B,F] join on D, etc.
+        // This can include external jobs
+        const joinObj = nextJobs.reduce((obj, jobName) => {
+            obj[jobName] = workflowParser.getSrcForJoin(workflowGraph, { jobName });
+
+            return obj;
+        }, {});
+
+        // Get pipelineId and job name from the `name`
+        // If internal, pipelineId will be the current pipelineId
+        const getPipelineAndJob = (name) => {
+            let pId;
+            let jName;
+
+            if (EXTERNAL_TRIGGER_AND.test(name)) {
+                [, pId, jName] = EXTERNAL_TRIGGER_AND.exec(name);
+            } else {
+                pId = pipelineId;
+                jName = name;
+            }
+
+            return { pId, jName };
+        };
+
+        console.log(joinObj);
+
+        // Go through each next job
+        return Promise.all(Object.keys(joinObj).map(async (nextJobName) => {
```

Comment:
Still need to handle restart case. For example
join list of `D` is `[A, sd@111:B, sd@222:C]`
`A and sd@222:C` passed but `sd@111:B` failed
if we restart `sd@111:B`, then `D` should know to fetch previous successful info from `A and sd@222:C` build. This will require additional logic